tests/unit/forklift/test_legacy.py

```diff
@@ -2310,31 +2310,53 @@ def test_upload_fails_with_diff_filename_same_blake2(
             "400 File already exists. See /the/help/url/ for more information."
         )
 
-    def test_upload_fails_with_wrong_filename(self, pyramid_config, db_request):
+    @pytest.mark.parametrize(
+        "filename, project_name",
+        [
+            # completely different
+            ("nope-{version}.tar.gz", "something_else"),
+            ("nope-{version}-py3-none-any.whl", "something_else"),
+            # starts with same prefix
+            ("nope-{version}.tar.gz", "no"),
+            ("nope-{version}-py3-none-any.whl", "no"),
+            # starts with same prefix with hyphen
+            ("no-way-{version}.tar.gz", "no"),
+            ("no_way-{version}-py3-none-any.whl", "no"),
+        ],
+    )
+    def test_upload_fails_with_wrong_filename(
+        self, pyramid_config, db_request, metrics, filename, project_name
+    ):
         user = UserFactory.create()
         pyramid_config.testing_securitypolicy(identity=user)
         db_request.user = user
+        db_request.user_agent = "warehouse-tests/6.6.6"
         EmailFactory.create(user=user)
-        project = ProjectFactory.create()
+        project = ProjectFactory.create(name=project_name)
         release = ReleaseFactory.create(project=project, version="1.0")
         RoleFactory.create(user=user, project=project)
 
-        filename = f"nope-{release.version}.tar.gz"
+        storage_service = pretend.stub(store=lambda path, filepath, meta: None)
+        db_request.find_service = lambda svc, name=None, context=None: {
+            IFileStorage: storage_service,
+            IMetricsService: metrics,
+        }.get(svc)
 
         db_request.POST = MultiDict(
             {
                 "metadata_version": "1.2",
                 "name": project.name,
                 "version": release.version,
                 "filetype": "sdist",
-                "md5_digest": "nope!",
+                "md5_digest": _TAR_GZ_PKG_MD5,
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏now that we're paramtrizing .whl files in the test case, does the same TAR_GZ behavior name apply to those files as well? Or is this 100% a testing-only behavior and we're unlikely to hit a whl-specific condition on prod? I suspect it's fine, but wanted to ask.

Reply:
The file contents don't really matter here as long as the hashes match, but I agree this could be a little confusing to future travelers.